src/line_launcher/mod.rs

```diff
@@ -112,149 +146,292 @@ impl LineLauncher {
+                    PlayingState::Playing {
+                        next_note_off_index,
+                        line_index,
+                        next_note_index,
+                        pitch_offset,
+                    } => {
+                        let line = &self.lines[line_index];
+                        let mut updated_next_note_off_index: Option<usize> = None;
+                        debug!("Received fire note on message: {:?}", fire_note_on_message);
+                        if fire_note_on_message.note_index > 0
+                            && next_note_off_index < fire_note_on_message.note_index
+                        {
+                            debug!(
+                                "Synchronously firing note off's, next_note_off_index: {}",
+                                next_note_off_index
+                            );
+                            for note_index in next_note_off_index..fire_note_on_message.note_index {
+                                let note = &line.notes[note_index];
+                                midi_message_sender
+                                    .fire_note_off(note.note.step(pitch_offset).unwrap());
+                                updated_next_note_off_index = Some(note_index + 1);
+                            }
+                        }
+                        let use_next_note_off_index =
+                            updated_next_note_off_index.unwrap_or(next_note_off_index);
+                        let note = &line.notes[fire_note_on_message.note_index];
+                        let note_with_offset = note.note.step(pitch_offset).unwrap();
+                        midi_message_sender.fire_note_on(note_with_offset);
+                        if let Some(duration_ratio) = duration_ratio {
+                            if let Some(duration_between_sixteenth_notes) =
+                                duration_between_sixteenth_notes.get_duration()
+                            {
+                                debug!(
+                                    "Sending schedule note off message, note_index: {}, now: {:?}, duration_ratio: {}",
+                                    fire_note_on_message.note_index, SystemTime::now(), duration_ratio
+                                );
+                                schedule_note_off_sender
+                                    .send(ScheduleNoteOffMessage {
+                                        time: SystemTime::now()
+                                            + duration_between_sixteenth_notes
+                                                .mul_f64(duration_ratio),
+                                        note_index: fire_note_on_message.note_index,
+                                    })
+                                    .unwrap();
+                            }
+                        }
+                        playing_state = if use_next_note_off_index >= line.notes.len() {
```

Comment:
This is duplicating the playing-state-transition logic above (the asynchronously-fired-note-off case vs this fired-note-off-synchronously-with-next-note-on case) and seems like it could be DRY'ed up into a more declarative expression of the state transition (eg `playing_state.advance_next_note_off_index(use_next_note_off_index)`)?